File: csharp_solutions/easy/ValidParens.cs
namespace ValidParens
{
    class Solution
    {
        public bool IsValid(string s)
        {
            var parenPairMap = new Dictionary<char, char>()
            {
                { ')', '(' },
                { ']', '[' },
                { '}', '{' },
            };
            var openStack = new Stack<char>();
            foreach (char c in s)
            {
                if (parenPairMap.ContainsKey(c))
                {
                    if (openStack.Count == 0)
                    {
                        return false;
                    }
                    char previousOpen = openStack.Pop();
                    if (parenPairMap[c] != previousOpen)
                    {
                        return false;
                    }
                }
                else
                {
                    openStack.Push(c);
                }
            }
            return openStack.Count == 0;
        }
    }
}

File: python_solutions/easy/valid_parens.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairMap = {
            "(": ")",
            "[": "]",
            "{": "}",
        }
        stack = []

        for c in s:
            if c in pairMap:
                stack.append(c)
            else:
                if not stack or pairMap[stack.pop()] != c:
                    return False
        return not stack
